inference_overview.md

```diff
@@ -0,0 +1,449 @@
+1. To make a prediction during a UDF, combine private embeddings from the
+  inference service running in a TEE with non-private embeddings from UDF
+  function arguments or the ad retrieval server store with an operation like a
+  dot product. This is the final prediction.
+
+The following figure shows a factorized pCTR model broken into three parts,
+contextual, user and ads towers. Only the User tower needs to run inside the
+TEE. The contextual and ad towers can run outside the TEE since they do not rely
+on private features.
+
+<figure id = "image-1">
+  <img src = "images/inference_overview_fig1.png"
+  width = "100%"
+  alt = "Factorized pCTR model">
+  <figcaption><b>Figure 1.</b> Factorized pCTR Model </figcaption>
+</figure>
+
+Factorized models are useful because loading and executing large models inside
+the TEE may be expensive. Oftentimes, only a small part of the model relies on
+private data (like user data), and the rest needs non-private (e.g. contextual
+or ad) data. With factorization, parts that depend on non-private features can
+be outside the TEE where they can be executed on ad tech hardware, while keeping
+the models requiring private features inside the TEE.
+
+Factorization can be utilized only when the model can be broken down into its
+constituents (as described above) fairly easily. In case models rely on a lot of
+cross features - for instance ad x contextual information, this may not be an
+effective strategy.
+
+We will support data flows that allow querying and passing such embeddings to
+`generateBid()`, where they can be combined with user embeddings to make the
+final prediction. The next section describes this data flow.
+
+## Overall flow
+
+### Loading external models
+
+B&A will support loading models inside the TEE. B&A servers will fetch and load
+ad tech models periodically from a connected cloud bucket and make them
+available for inference in the ad tech provided code (e.g. `generateBid()`). The
+loading will be performed in the background, separate from the execution of
+`generateBid()`. After loading, the models will also be pre-warmed in the
+background, so that when the models are used in the request flow, they are
+already warmed.
+
+<figure id = "image-2">
+  <img src = "images/inference_overview_fig2.png"
+  width = "100%"
+  alt = "Loading external models">
+  <figcaption><b>Figure 2.</b> Loading external models</figcaption>
+</figure>
+
+The loaded models will be executed in the supported frameworks (e.g. Tensorflow,
+PyTorch). When `runInference()` is called from within `generateBid()`, the call
+will be forwarded to the corresponding executor to execute the inference
+request, and return the results back to the calling function. These calls occur
+within the confines of a single TEE host, and data never leaves the TEE.
+
+Note that `runInference()will` not call out to the cloud buckets, or trigger
+loading of the model. It will only execute requests against models that have
```

Comment:
Nit: "against the pre-loaded models."

Reply:
I think the current wording is clearer?

Reply:
Resolving. I didn't want to use the word pre-loaded, since it sounds like the model might be pre-loaded with the code or something. Let me know what you think!